li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/largemessage/ConsumerRecordsProcessor.java

```diff
@@ -172,12 +173,19 @@ public Long safeOffset(TopicPartition tp, long messageOffset) {
       if (earliestTrackedOffset != null && origOffset > earliestTrackedOffset) {
         // We need to find the previous delivered offset from this partition and use its safe offset.
         Long previousDeliveredOffset = _deliveredMessageOffsetTracker.closestDeliveredUpTo(tp, origOffset - 1);
-        if (previousDeliveredOffset != null) {
-          safeOffsetToCommit = _deliveredMessageOffsetTracker.safeOffset(tp, previousDeliveredOffset);
+        Long latestDelivered = _deliveredMessageOffsetTracker.delivered(tp); //latest ever delivered for this partition
+
+        if (Objects.equals(previousDeliveredOffset, latestDelivered)) {
+          //user is attempting to commit "current location" (as opposed to arbitrary seek shenanigans)
+          //in this case we must also take into account the earliest offset for which we still have
+          //any hope of delivering a large message (maybe we dropped a lot of them due to low memory)
+          safeOffsetToCommit = _messageAssembler.safeOffset(tp);
         } else {
-          //In this case nothing has been delivered to the consumer (this this partition) either because we never
-          // reached the highwatermark or because we were waiting to assemble some message
-          safeOffsetToCommit = _deliveredMessageOffsetTracker.safeOffset(tp);
+          if (previousDeliveredOffset != null) {
+            safeOffsetToCommit = _deliveredMessageOffsetTracker.safeOffset(tp, previousDeliveredOffset);
+          } else {
+            safeOffsetToCommit = _messageAssembler.safeOffset(tp);
```

Comment:
What does this do?

Reply:
deliveredMessageOffsetTracker has history, so can map "historical" offsets delivered to users to their "real" underlying offsets.

messageAssembler only deals with the current stats.

this code means that if we have history (because we've delivered some != 0 msgs before, we use deliveredMessageOffsetTracker to get the real offset, otherwise use messageAssembler